distributed/worker.py

```diff
@@ -1718,7 +1844,19 @@ def transition_ready_waiting(self, ts):
         """
         This transition is common for work stealing
         """
-        pass
+        ts.runspec = None
+
+    def transition_waiting_new(self, ts):
+        """
+        Common in work stealing
+        """
+        ts.runspec = None
+
+    def transition_constrained_waiting(self, ts):
```

Comment:
I'm not sure if this is possible. Constrained is more or less the twin state of a `waiting` task with the distinction that the constrained task has resource constraints to fulfil. If that's the case, this transition would mean we're changing the definition of the task by stripping the resource constraints

Reply:
This should only be called if constrained task is stolen -- the current logic is there is that stolen tasks revert to a state of `waiting` and are then released. 
i think in this case we're only stripping the resource constraints because they're being enacted on a different worker.

Reply:
Even if it is stolen, the resource constraint of a task is defined on task initialisation, effectively by the user. ressource counting is done by the worker but the task should not be modfiied

it is managed here
https://github.com/dask/distributed/blob/98570fbcd662bc0fe468ccba9d57d582d2a3c678/distributed/worker.py#L1725-L1726
and here
https://github.com/dask/distributed/blob/98570fbcd662bc0fe468ccba9d57d582d2a3c678/distributed/worker.py#L1740-L1741
but I would consider the field `resource_restrictions` to be immutable (if only that was easily possible in python :) )

Therefore, if my reasoning is correct, this might overshadow another wrong transition (maybe key is forgotten and improperly resubmitted?)

Reply:
>the current logic is there is that stolen tasks revert to a state of waiting and are then released.

Do you mean this line?

https://github.com/dask/distributed/blob/98570fbcd662bc0fe468ccba9d57d582d2a3c678/distributed/worker.py#L2285

I think it is wrong... :( This should show up somewhere in a verify task method. I think there should be something like

```python
def validate_task_waiting(self, ts):
    ...
    assert ts.resource_restrictions is None # // or empty dict, dont know
    ...
```

Reply:
That's a very good point re: `resource_restrictions` -- I'll look again at the stealing code.
I don't know why the transition to `waiting` was in there -- it seems like after we have confirmation that the task has been stolen, we should release the key on the victim side (along with whatever `available_resources` might have been allocated for use)

Reply:
Also, it looks like `steal_request` doesn't need to worry about releasing `available_resources` because they aren't "taken" until the task transitions from `ready -> executing`, at which point it's no longer eligible to be stolen